### PR TITLE
Merge 3.3

### PR DESCRIPTION
**Pull request by wallyworld**
Merge 3.3

https://github.com/juju/juju/pull/16732 [from wallyworld/fix-cache-race](https://github.com/juju/juju/commit/4acafa2720489055203414a28ae6bb2e6b2de67d)

### CODE REVIEW

File: core/cache/resident.go [hunk 1]
```diff
@@ -229,6 +230,17 @@ func (r *Resident) CacheId() uint64 {
 func (r *Resident) registerWorker(w worker.Worker) func() {
 	id := r.nextResourceId()
 	r.mu.Lock()
+	// If this resident is already being evicted
+	// don't register any new workers.
+	if r.evicting {
+		return func() {
```
Comment (SimonRichardson) [inline core/cache/resident.go:236]: You're missing an `Unlock` here, as you're holding for the `r.evicting`, but not releasing after the if.

File: tests/includes/juju.sh [hunk 2]
```diff
@@ -350,7 +350,7 @@ post_add_model() {
 	model=${2}
 
 	ctrl_arg="${controller}:${model}"
-	log_file="${controller}-${model}.log"
+	log_file="${controller}-${model}-debug.log"
```
Comment (SimonRichardson) [inline tests/includes/juju.sh:353]: Thanks.

File: tests/suites/model/migration.sh [hunk 3]
```diff
@@ -6,7 +6,9 @@ run_model_migration() {
 	# Ensure we have another controller available.
 	bootstrap_alt_controller "alt-model-migration"
 	juju switch "alt-model-migration"
-	juju add-model "model-migration"
+	add_model "model-migration"
+	juju model-config -m controller "logging-config=#migration=DEBUG"
+	juju model-config -m model-migration "logging-config=#migration=DEBUG"
```
Comment (SimonRichardson) [inline tests/suites/model/migration.sh:11]: Nice, this makes life a lot easier.

**Review by SimonRichardson — CHANGES_REQUESTED** [review]

File: core/cache/resident.go (hunk 1 above)
Comment (wallyworld) [reply core/cache/resident.go:237]: Ooops, ty, Fixed.

**Review by SimonRichardson — APPROVED** [review]

**Pull request merged**